### PR TITLE
[feature](mysql) Support secure MySQL connection to FE

### LINKED ISSUE #16313: [Feature] Support secure MySQL connection to FE
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Doris currently does not support SSL connection from MySQL clients, it's not secure enough in some cases, especially access Doris via the public internet.

image

### Use case

Add SSL support in MySQL protocol
image

image

### Related issues

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Yukang-Lian**
# Proposed changes

Issue Number: close #16313 

## Problem summary

Background:
Doris currently does not support SSL connection from MySQL clients, it's not secure enough in some cases, especially access Doris via the public internet.

Solution:
- Use TLS1.2 protocol to encrypt information.
- Implementation details
  * server <--- connect <--- client
  * if enable SSL: {
  * server <--- SSL connection request packet <--- client
  * --------- SSL Exchange --------- } (we will add this `if` logic part in this PR)
  * server --- handshake request packet ---> client
- reference1 https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase.html#sect_protocol_connection_phase_initial_handshake_ssl_handshake
- reference2 https://www.rfc-editor.org/rfc/rfc5246

close #16313

Signed-off-by: Yukang Lian <yukang.lian2022@gmail.com>
Co-authored-by: Gavin Chou <gavineaglechou@gmail.com>

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.77 seconds
 load time: 504 seconds
 storage size: 17123080400 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230201205615_clickbench_pr_89093.html

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlCapability.java [hunk 1]
```diff
@@ -75,7 +75,14 @@ public String toString() {
     private static final int DEFAULT_FLAGS = Flag.CLIENT_PROTOCOL_41.getFlagBit()
             | Flag.CLIENT_CONNECT_WITH_DB.getFlagBit() | Flag.CLIENT_SECURE_CONNECTION.getFlagBit()
             | Flag.CLIENT_PLUGIN_AUTH.getFlagBit() | Flag.CLIENT_LOCAL_FILES.getFlagBit();
+
+    private static final int SSL_FLAGS = Flag.CLIENT_PROTOCOL_41.getFlagBit()
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlCapability.java:79]: Flag.CLIENT_LOCAL_FILES.getFlagBit() | Config.enable_ssl ? Falg : 1;

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlProto.java [hunk 2]
```diff
@@ -45,6 +45,7 @@
 // MySQL protocol util
 public class MysqlProto {
     private static final Logger LOG = LogManager.getLogger(MysqlProto.class);
+    public static final boolean USE_SSL = Config.enable_ssl;
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlProto.java:48]: Redundant, Config.enable_ssl is enough.

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlProto.java [hunk 3]
```diff
@@ -170,8 +171,44 @@ public static boolean negotiate(ConnectContext context) throws IOException {
...
+
+        // Server receive SSL connection request packet from client.
+        ByteBuffer sslConnectionRequest;
         // Server receive authenticate packet from client.
-        ByteBuffer handshakeResponse = channel.fetchOnePacket();
+        ByteBuffer handshakeResponse;
+
+        if (capability.isSsl()) {
+            // During development, we set SSL mode to true by default
+            if (USE_SSL) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlProto.java:186]: Config.enable_ssl

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslExchange.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.mysql;
+
+public class MysqlSslExchange {
+    public static final boolean sslExchange() {
+        // TODO: TLS protocol
+        return true;
+    }
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslExchange.java:24]: SSLEngine inside.

class ConnectContext {
...
MysqlSslExchange sslExchange;
...
}:

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslConnectionContext.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+import org.apache.logging.log4j.Logger;
+
+import java.nio.ByteBuffer;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+
+public class MysqlSslConnectionContext {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslConnectionContext.java:30]: MysqlSslContext

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlProto.java [hunk 6]
```diff
@@ -170,8 +173,49 @@ public static boolean negotiate(ConnectContext context) throws IOException {
...
+                sslConnectionRequest = clientRequestPacket;
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlProto.java:189]:
```
mysqlChannel.isSslMode = true;

boolean isSslMode() { return isSslMode; }
```

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlChannel.java [hunk 7]
```diff
@@ -197,14 +285,42 @@ protected void realNetSend(ByteBuffer buffer) throws IOException {
...
+            dstBuf.put(netData);
+            dstBuf.flip();
+        }
+    }
+
     public void flush() throws IOException {
-        if (null == sendBuffer || sendBuffer.position() == 0) {
+        ByteBuffer sendData ;
+        if(isSslMode|isHandshaking){
+            sendData = sendSslBuffer;
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlChannel.java:302]: Use ternary expression to reduce lines of code, e.g.
```
sendData = x ? a : b;
```

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlHandshakePacket.java [hunk 8]
```diff
@@ -25,11 +25,12 @@ public class MysqlHandshakePacket extends MysqlPacket {
     // JDBC uses this version to check which protocol the server support
     // Set the patch version to 99 to prevent the vulnerability scanning tool from
     // falsely reporting MySQL vulnerabilities
-    public static final String SERVER_VERSION = "5.7.99";
+    public static final String SERVER_VERSION = "8.0.31";
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlHandshakePacket.java:28]: This version may not be updated due to TLS does not complete all the features of 8.0.31.

File: fe/fe-common/src/main/java/org/apache/doris/common/Config.java [hunk 9]
```diff
@@ -1969,6 +1969,12 @@ public class Config extends ConfigBase {
     @ConfField(masterOnly = true, mutable = true)
     public static int max_error_tablet_of_broker_load = 3;
 
+    /**
+     * If set to ture, doris will establish an encrypted channel based on the SSL protocol with mysql.
+     */
+    @ConfField(mutable = false, masterOnly = false)
+    public static boolean enable_ssl = true;
+
```
Comment (gavinchou) [inline fe/fe-common/src/main/java/org/apache/doris/common/Config.java:1977]: Should we add a conf filed for certificate?

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run p0

**Comment by Yukang-Lian** [comment]
run feut

**Comment by Yukang-Lian** [comment]
run feut

**Comment by Yukang-Lian** [comment]
run buildall

File: regression-test/conf/regression-conf.groovy [hunk 10]
```diff
@@ -24,7 +24,6 @@ defaultDb = "regression_test"
 // init cmd like: select @@session.tx_read_only
 // at each time we connect.
 // add allowLoadLocalInfile so that the jdbc can execute mysql load data from client.
-jdbcUrl = "jdbc:mysql://127.0.0.1:9030/?useLocalSessionState=true&allowLoadLocalInfile=true"
```
Comment (gavinchou) [inline regression-test/conf/regression-conf.groovy:27]: Regression test may change this line accordingly, which will skip testing tls if it is deleted.

File: regression-test/framework/src/main/groovy/org/apache/doris/regression/Config.groovy [hunk 11]
```diff
@@ -267,7 +268,15 @@ class Config {
 
         if (config.jdbcUrl == null) {
             //jdbcUrl needs parameter here. Refer to function: buildUrl(String dbName)
-            config.jdbcUrl = "jdbc:mysql://127.0.0.1:9030/?useLocalSessionState=true&allowLoadLocalInfile=true"
```
Comment (gavinchou) [inline regression-test/framework/src/main/groovy/org/apache/doris/regression/Config.groovy:270]: We also have to support the config which sets a jdbcUrl.

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslContext.java [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+public class MysqlSslContext {
+
+    private static final Logger LOG = LogManager.getLogger(MysqlSslContext.class);
+    private SSLEngine sslEngine;
+    private SSLContext sslContext;
+    private String protocol;
+    private ByteBuffer serverAppData;
+    private ByteBuffer serverNetData;
+    private static final String keyStoreFile = "../../../regression-test/certificate.p12";
+    private static final String trustStoreFile = "../../../regression-test/certificate.p12";
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslContext.java]: We should get the paths from Config.java -- make them configurable.

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslContext.java [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+        // init data buffer
+        initDataBuffer();
+        // set channel sslengine.
+        channel.setSslEngine(sslEngine);
+        // begin handshake.
+        sslEngine.beginHandshake();
+        while (sslEngine.getHandshakeStatus() != HandshakeStatus.FINISHED
+                && sslEngine.getHandshakeStatus() != HandshakeStatus.NOT_HANDSHAKING) {
+            //            if ((System.currentTimeMillis() - startTime) > 10000) {
+            //                throw new Exception("try to establish SSL connection failed, timeout!");
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslContext.java:138]: Remove if unused.

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslContext.java [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+        HandshakeStatus hsStatus = sslEngine.getHandshakeStatus();
+        if (hsStatus == HandshakeStatus.NEED_TASK) {
+            throw new Exception("handshake shouldn't need additional tasks");
+        }
+    }
+
+    private void handleNeedWrap(MysqlChannel channel) {
+        try {
+            serverAppData.flip();
+            serverAppData.clear();
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslContext.java:183]: It seems these 4 lines are redundant.

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: regression-test/suites/mysql_ssl_p0/test_mysql_connection.groovy [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+suite("test_mysql_connection") {
+    def sout = new StringBuilder(), serr = new StringBuilder()
+    'mysql -uroot -P9030 -h127.0.0.1'.execute()
```
Comment (gavinchou) [inline regression-test/suites/mysql_ssl_p0/test_mysql_connection.groovy:20]: Extract host and port from config.jdbcUrl
and add a command to exec
```
mysql -u .. -h ... -P...  -e "show databases"
```

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by gavinchou** [comment]
run compile

File: fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslContext.java [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+            KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+            kmf.init(ks, password);
+
+            TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+            tmf.init(ts);
+            sslContext = SSLContext.getInstance(protocol);
+            sslContext.init(kmf.getKeyManagers(), tmf.getTrustManagers(), null);
+        } catch (NoSuchAlgorithmException | KeyManagementException | KeyStoreException | IOException
+                 | CertificateException | UnrecoverableKeyException e) {
+            LOG.error("Failed to initialize SSL because", e);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/mysql/MysqlSslContext.java:85]: Do we need to exit the entire FE process if something is wrong (e.g. no such file)?
If not, can FE run without SSL support properly when it fails to init ssl context here?

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: build.sh [hunk 17]
```diff
@@ -305,7 +305,7 @@ if [[ -z "${DISABLE_JAVA_UDF}" ]]; then
 fi
 
 if [[ -z "${DISABLE_JAVA_CHECK_STYLE}" ]]; then
-    DISABLE_JAVA_CHECK_STYLE='OFF'
+    DISABLE_JAVA_CHECK_STYLE='ON'
```
Comment (gavinchou) [inline build.sh:308]: Donot change this.

File: build.sh [hunk 18]
```diff
@@ -479,10 +479,10 @@ function build_ui() {
 }
 
 # FE UI must be built before building FE
-if [[ "${BUILD_FE}" -eq 1 ]]; then
-    build_ui
-fi
-
+#if [[ "${BUILD_FE}" -eq 1 ]]; then
+#    build_ui
```
Comment (gavinchou) [inline build.sh:483]: Re-enable it.

File: regression-test/suites/mysql_ssl_p0/test_mysql_connection.groovy [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+suite("test_mysql_connection") {
+    """mysql -uroot -P9030 -h127.0.0.1 -e 'show databases; status;'""".execute()
```
Comment (gavinchou) [inline regression-test/suites/mysql_ssl_p0/test_mysql_connection.groovy:19]: We need to check the result.

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: build.sh (hunk 17 above)
Comment (Yukang-Lian) [reply build.sh:308]: this is just for test, will remove it if all test passed.

File: build.sh (hunk 18 above)
Comment (Yukang-Lian) [reply build.sh:483]: ditto

File: regression-test/suites/mysql_ssl_p0/test_mysql_connection.groovy (hunk 19 above)
Comment (Yukang-Lian) [reply regression-test/suites/mysql_ssl_p0/test_mysql_connection.groovy:19]: will do it soon.

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run p0

**Comment by Yukang-Lian** [comment]
this PR was refactored to #17138, so close this.

**Pull request closed without merging**